File: js/70ClimbingStairs.js
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function (x) {
  const m = { 1: 1, 2: 2 };
  function f(n, m) {
    if (m[n]) return m[n];
    let i = 3;
    while (i <= x) {
      m[i] = f(i - 1, m) + f(i - 2, m);
      i++;
    }
    return m[n];
  }
  return f(x, m);
};
// https://leetcode.com/problems/climbing-stairs/discuss/25299/Basically-it's-a-fibonacci.

File: swift/70ClimbingStairs.swift
// https://leetcode.com/problems/climbing-stairs/solutions/2383701/fastest-solution-explained-0ms-100-o-n-time-complexity-o-n-space-complexity/
class Solution {
    // - Complexity:
    //   - time: O(n), where n is the number of steps.
    //   - space: O(1), only constant space is used.

    func climbStairs(_ n: Int) -> Int {
        if n == 0 || n == 1 { return 1 }

        var first = 1
        var second = 1
        var third = 0

        for _ in 2...n {
            third = first + second
            first = second
            second = third
        }

        return second
    }

}